docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,389 @@
+
+![relationships of ipam custom resources](images/ipam-integration/consumption.png)
+
+More implementation details can be found [here](#consuming-as-an-infrastructure-provider).
+
+#### Example
+
+Using CAPV as an example, a template *could* look like this:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+ name: example
+ namespace: vsphere-site1
+spec:
+ template:
+   spec:
+     cloneMode: FullClone
+     numCPUs: 8
+     memoryMiB: 8192
+     diskGiB: 45
+     network:
+       devices:
+       - dhcp4: false
+         fromPool: # reference to the pool
+           group: ipam.cluster.x-k8s.io/v1alpha1
+           kind: IPPool
+           name: testpool
+```
+
+A Machine generated from that template could then look like this:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachine
+metadata:
+name: example-1
+namespace: vsphere-site1
+spec:
+ cloneMode: FullClone
+ numCPUs: 8
+ memoryMiB: 8192
+ diskGiB: 45
+ network:
+   devices:
+   - dhcp4: false
+     fromPool: # reference to the pool
+       group: ipam.cluster.x-k8s.io/v1alpha1
+       kind: IPPool
+       name: testpool
+status:
+ network:
+   devices:
+     - claim:
+         name: example-1-1
+```
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator I want to use static IP addresses with my Machines to avoid problems with DHCP lease duration during rolling cluster upgrades.
+
+#### Story 2
+
+As a cluster operator I want to integrate machines provisioned by CAPI with an external IPAM system and allocate IP addresses from it.
+
+#### Story 3
+
+As an infrastructure provider I want a single interface to integrate with multiple different IPAM providers.
+
+#### Story 4
+
+As an IPAM provider I want a single interface to integrate with all infrastructure providers that can benefit from IPAM integration.
+
+### Implementation Details/Notes/Constraints
+
+#### New API Types
+
+The following API Types should be added to cluster-api. In the first iteration they will be added to the experimental API and will therefore live in the `cluster.x-k8s.io` group.
+
+##### IPAddressClaim
+
+The IPAddressClaim is used to request ip addresses from a pool. It gets reconciled by IPAM providers, which can filter based on the `Spec.Pool` reference.
+
+After an IPAddressClaim is created, the `Spec.Pool` reference, and therefore essentially the entire `Spec` is immutable to simplify reconciliation of the resource.
+
+```go
+// IPAddressClaimSpec describes the desired state of an IPAddressClaim
+type IPAddressClaimSpec struct {
+  // Pool is a reference to the pool from which an IP address should be allocated.
+  Pool LocalObjectReference `json:"pool,omitempty"`
+}
+
+// IPAddressClaimStatus contains the status of an IPAddressClaim
+type IPAddressClaimStatus struct {
+  // Address is a reference to the address that was allocated for this claim.
+  Address LocalObjectReference `json:"address,omitempty"`
```

Comment:
Is it a problem if we lose this information through a clusterctl move?

Reply:
If ipam providers adhere to the principle of always being able to recompute status - no.

Reply:
Ah I missed that the IPAddress has a reference to the Claim, so it's relatively easy to just set it again.

Reply:
Also the name of the claim and the address is probably going to be identical (or the providers have to figure out another way of how to correlate them).

Reply:
We just have to be clear if we want to depend on the name being the same (there was some fun regarding Machine and InfraMachine names in the past)

Reply:
I'd say we won't, that's up to the provider.